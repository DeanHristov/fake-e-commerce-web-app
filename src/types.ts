export enum PRODUCT_CATEGORIES {
  ELECTRONIC = 'ELECTRONIC',
}

export type TCurrency = 'USD' | 'BGN' | 'EUR';

export interface IComment {
  productId: string;
  _id: string;
  name: string;
  email: string;
  body: string;
  date: string;
}

// Object schema can be seen here: https://dummyjson.com/docs/products
export interface IProduct {
  id: number;
  title: string;
  description: string;
  price: number;
  discountPercentage: number;
  rating: number;
  stock: number;
  brand: string;
  category: string;
  thumbnail: string;
  images?: string[];
  comments?: IComment[];
}

export interface IUser {
  name: string;
  email: string;
  password: string;
  isAdmin: boolean;
}

export interface ICartProduct {
  id: number;
  price: number;
  quantity: number;
  discountPercentage?: number;
  // TODO Fix next line!
  discountedPrice?: number;
  title: string;
  thumbnail: string;
  total: number;
  inStock: number;
}

export interface ICart {
  id: number;
  userId: number;
  total: number;
  discountedTotal: number;
  totalProducts: number;
  totalQuantity: number;
  currency?: TCurrency;
  products: Record<number, ICartProduct>; // Map<number, ICartProduct>;
}

//@see:https://nextjs.org/docs/app/api-reference/functions/generate-metadata#generatemetadata-function
export interface IMetadataPageTitle {
  title: string;
}

export interface IProductsResponse {
  products: IProduct[];
  total: number;
  skip: number;
  limit: number;
}

export interface ICartsResponse {
  carts: ICart[];
  total: number;
  skip: number;
  limit: number;
}
