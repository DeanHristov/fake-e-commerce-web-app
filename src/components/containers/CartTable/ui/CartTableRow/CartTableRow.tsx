import { FC } from 'react';
import Image from 'next/image';
import { ICartProduct, TCurrency } from '@/types';

import Quantity from '@/components/ui/Quantity';
import PriceLabel from '@/components/ui/PriceLabel';
import ProductHeading from '@/components/ui/ProductHeading';
import RowControls from '@/components/containers/CartTable/ui/RowControls/RowControls';

export interface ICartTableRowProps extends ICartProduct {
  currency?: TCurrency;
}

const CartTableRow: FC<ICartTableRowProps> = ({
  id,
  price,
  quantity,
  discountedPrice,
  thumbnail,
  title,
  discountPercentage,
  currency = 'USD',
}) => {
  return (
    <ul className="w-full h-auto border-b border-b-dark px-2 py-2 md:flex md:justify-start">
      <li className="w-full md:flex md:justify-start md:space-x-2 overflow-hidden">
        <Image
          priority
          src={thumbnail}
          alt={title}
          width={0}
          height={0}
          sizes="100vw"
          className={`w-auto h-auto rounded-lg min-w-[4rem] max-w-[8rem] border border-gray-300 mx-auto md:m-0`}
        />
        <ProductHeading
          id={id}
          title={title}
          className={'text-center md:text-left'}
        >
          <span className="text-black text-sm">{`${quantity}x `}</span>
        </ProductHeading>
      </li>
      <li className="w-full flex justify-center space-x-2 md:pt-2 md:justify-start md:max-w-[8rem]">
        <h3 className="md:hidden">Price: </h3>
        <PriceLabel currency={currency} price={price} />
      </li>
      <li className="w-full md:max-w-[8rem] px-2 pt-2 mx-auto md:p-0">
        <Quantity
          onIncrease={(value) => console.log('onIncrease to: ' + value)}
          onDecrease={(value) => console.log('onDecrease to: ' + value)}
          initialValue={quantity}
        />
      </li>
      <li className="w-full md:max-w-[10rem] relative">
        <div className="w-full flex items-center justify-center space-x-2 py-2 md:py-0 md:flex-col">
          <h3 className="md:hidden">Subtotal: </h3>
          <PriceLabel
            currency={currency}
            discountPercentage={discountPercentage}
            price={discountedPrice}
          />
          <span className="text-red-500 font-medium text-center block">{`(-${discountPercentage}%)`}</span>
        </div>
        <RowControls
          className="text-center md:text-left md:absolute md:bottom-0.5 md:right-0"
          onRemove={() => console.log('item is removed')}
          onWishList={() => console.log('moved to wish list')}
        />
      </li>
    </ul>
  );
};

export default CartTableRow;
