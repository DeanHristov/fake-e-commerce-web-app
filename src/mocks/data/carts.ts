import { ICart } from '@/types';

export const carts: ICart[] = [
  {
    _id: '34e1d41b-384a-4bf4-866e-ab5b268d124c',
    products: {
      '8f12eb53-18c8-46a6-b975-ea0ee8b62d4c': {
        _id: '8f12eb53-18c8-46a6-b975-ea0ee8b62d4c',
        title: 'Spring and summershoes',
        price: 20,
        quantity: 3,
        total: 60,
        discountPercentage: 8.71,
        discountedPrice: 55,
        thumbnail: 'https://i.dummyjson.com/data/products/59/thumbnail.jpg',
        inStock: 5,
      },
    },
    currency: 'EUR',
    total: 2,
    discountedTotal: 1941,
    userId: 'ba500896-80ca-42bd-b7f7-a743b959203e',
    totalProducts: 2,
    totalQuantity: 2,
  },
];
