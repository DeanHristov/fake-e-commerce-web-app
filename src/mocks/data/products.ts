import { IProduct, PRODUCT_CATEGORIES } from '../../types';

const products: IProduct[] = [
  {
    _id: '29a34b48-8343-4e3d-8c21-3548ba249936',
    name: 'Airpods Pro2 Headphones',
    image: '/images/airpods.jpg',
    description:
      "Apple's new H2 chip is the driving force behind AirPods Pro and their improved sound characteristics. It works in unison with a personalized driver and amplifier for crystal clear heights and deep, rich bass with incredible clarity, so every sound is more pronounced than ever",
    brand: 'Apple',
    category: PRODUCT_CATEGORIES.ELECTRONIC,
    price: 399.99,
    countInStock: 10,
    rating: 4.5,
    currency: 'USD',
    reviews: 5,
  },
];

export default products;
