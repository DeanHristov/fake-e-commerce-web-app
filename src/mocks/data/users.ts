import bcrypt from 'bcryptjs';
import { IUser } from '../../types';

const users: IUser[] = [
  {
    name: 'Admin User',
    email: 'admin@example.com',
    password: bcrypt.hashSync('123456', 10),
    isAdmin: true,
  },
  {
    name: 'Regular User',
    email: 'regular-user@example.com',
    password: bcrypt.hashSync('123456', 10),
    isAdmin: false,
  },
];

export default users;
