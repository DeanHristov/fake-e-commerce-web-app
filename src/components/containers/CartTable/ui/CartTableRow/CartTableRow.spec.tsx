import { render } from '@testing-library/react';

import { carts } from '@/mocks/data/carts';
import { ICartProduct } from '@/types';
import CartTableRow from './CartTableRow';

const firstCardProduct: ICartProduct =
  carts[0].products['8f12eb53-18c8-46a6-b975-ea0ee8b62d4c'];

describe('UI/CartTableRow <CartTableRow {...} />', () => {
  it('Should render a default state of the Component', () => {
    const { container } = render(<CartTableRow {...firstCardProduct} />);

    expect(container).toMatchSnapshot();
  });

  it('Should render a default state of the Component', () => {
    const { container } = render(<CartTableRow {...firstCardProduct} />);

    expect(container).toMatchSnapshot();
  });
});
