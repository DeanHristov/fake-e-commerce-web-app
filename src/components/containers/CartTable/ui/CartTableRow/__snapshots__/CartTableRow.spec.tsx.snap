// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI/CartTableRow <CartTableRow {...} /> Should render a default state of the Component 1`] = `
<div>
  <ul
    class="w-full h-auto border-b border-b-dark px-2 py-2 md:flex md:justify-start"
  >
    <li
      class="w-full md:flex md:justify-start md:space-x-2 overflow-hidden"
    >
      <img
        alt="Spring and summershoes"
        class="w-auto h-auto rounded-lg min-w-[4rem] max-w-[8rem] border border-gray-300 mx-auto md:m-0"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="0"
        sizes="100vw"
        src="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=3840&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=3840&q=75 3840w"
        style="color: transparent;"
        width="0"
      />
      <h3
        class="text-black text-lg mt-2 truncate hover:underline hover:cursor-pointer text-center md:text-left"
      >
        <span
          class="text-black text-sm"
        >
          3x 
        </span>
        Spring and summershoes
      </h3>
    </li>
    <li
      class="w-full flex justify-center space-x-2 md:justify-start md:max-w-[8rem]"
    >
      <h3
        class="md:hidden"
      >
        Price: 
      </h3>
      <div
        class=""
      >
        <span
          class="price-label"
        >
          $20.00
        </span>
      </div>
    </li>
    <li
      class="w-2/3 px-2 pt-2 mx-auto md:w-full md:max-w-[8rem] md:p-0"
    >
      <div
        class="flex justify-center items-center bg-gray-100 rounded-md border border-gray-400 undefined"
      >
        <button
          class="text-xl w-20 text-red-400 font-medium border-r border-r-gray-400"
        >
          -
        </button>
        <span
          class="w-full px-1  text-center"
        >
          3
        </span>
        <button
          class="text-xl w-20  text-green-400 font-medium border-l border-l-gray-400"
        >
          +
        </button>
      </div>
    </li>
    <li
      class="w-full md:max-w-[10rem] relative"
    >
      <div
        class="w-full flex items-center justify-center space-x-2 py-2 md:py-0 md:flex-col"
      >
        <h3
          class="md:hidden"
        >
          Subtotal: 
        </h3>
        <div
          class="space-x-2"
        >
          <span
            class="price-label-old line-through text-sm"
          >
            $59.79
          </span>
          <span
            class="price-label-old"
          >
            /
          </span>
          <span
            class="price-label"
          >
            $55.00
          </span>
        </div>
        <span
          class="text-red-500 font-medium text-center block"
        >
          (-8.71%)
        </span>
      </div>
      <div
        class="text-black text-sm text-center md:text-left md:absolute md:bottom-0.5 md:right-0"
      >
        <span
          class="hover:underline hover:cursor-pointer hover:text-dark"
          data-testid="btn-remove"
        >
          Remove
        </span>
         / 
        <span
          class="hover:underline hover:cursor-pointer hover:text-dark"
          data-testid="btn-wish-list"
        >
          Wish list
        </span>
      </div>
    </li>
  </ul>
</div>
`;
